lib/Dialect/TritonIntelGPU/Transforms/MatchTargetSize.cpp

```diff
@@ -0,0 +1,607 @@
+  std::tuple<SmallVector<int64_t>, Type, SmallVector<int64_t>>
+  getSubTypeAndShape(Type type) const;
+
+  void transformMakeTensorPtrOp(tt::MakeTensorPtrOp op);
+  void transformArithConstantOp(arith::ConstantOp op);
+  void transformDotOp(tt::DotOp dot);
+  void transformGenericOp(Operation *op);
+
+  DenseMap<Attribute, SmallVector<int64_t>> sizePerAttrMap;
+  DenseSet<Attribute> dotAttrs;
+};
+
+/// Simplify arith operations with constant RHS.
+class ArithRemPattern : public OpRewritePattern<arith::RemSIOp> {
+public:
+  using OpRewritePattern<arith::RemSIOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(arith::RemSIOp op,
+                                PatternRewriter &rewriter) const final {
+    APInt value;
+    if (!matchPattern(op.getRhs(), m_ConstantInt(&value)))
+      return failure();
+
+    Location loc = op.getLoc();
+    Type type = op.getType();
+
+    if (value.isOne()) {
+      IntegerAttr attr = rewriter.getIntegerAttr(type, 0);
+      auto zero = rewriter.create<arith::ConstantOp>(loc, attr);
+      rewriter.replaceOp(op, zero);
+      return success();
+    }
+
+    if (value.popcount() == 1) {
+      IntegerAttr attr =
+          rewriter.getIntegerAttr(type, value.getSExtValue() - 1);
+      auto mask = rewriter.create<arith::ConstantOp>(loc, attr);
+      auto result = rewriter.create<arith::AndIOp>(loc, op.getLhs(), mask);
+      rewriter.replaceOp(op, result);
+      return success();
+    }
+
+    return failure();
+  }
+};
+
+/// Simplify extract operations.
+/// FIXME: this assume that there is not sideEffect op in between.
+class ExtractPattern : public OpRewritePattern<ttgi::ExtractOp> {
+public:
+  using OpRewritePattern<ttgi::ExtractOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(ttgi::ExtractOp op,
+                                PatternRewriter &rewriter) const final {
+    Value base = op.getBase();
+    if (Operation *def = base.getDefiningOp()) {
+      if (auto glue = dyn_cast<ttgi::GlueOp>(def)) {
+        Value sub = glue->getOperand(op.getIndex());
+        rewriter.replaceOp(op, sub);
+        return success();
+      }
+    }
+
+    if (base.getType() == op.getType() && op.getIndex() == 0) {
+      rewriter.replaceOp(op, base);
+      return success();
+    }
+
+    return failure();
+  }
+};
+
+/// Simplify SCF loops.
+class ScfPattern : public OpRewritePattern<scf::ForOp> {
+public:
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(scf::ForOp op,
+                                PatternRewriter &rewriter) const final {
+    SmallVector<Operation *> deleteList;
+    SmallVector<Value> newInits;
+    DenseMap<Value, int> userIndexMap;
+    unsigned idx = 0;
+    for (auto [arg, init] : llvm::zip(op.getRegionIterArgs(), op.getInits())) {
+      auto glue = dyn_cast<ttgi::GlueOp>(init.getDefiningOp());
+      if (!glue) {
+        newInits.push_back(init);
+        userIndexMap[arg] = idx;
+        idx++;
```

Comment:
https://llvm.org/docs/CodingStandards.html#prefer-preincrement
```suggestion
        ++idx;
```

Reply:
Same goes to other increment.

Reply:
done